### PR TITLE
Fix: protect against cyclic scopes

### PR DESCRIPTION
**Pull request by hsheth2**
This ensures that we don't hit infinite loops.

Given that this is a defensive measure, I'm not sure what the best way to test this would be. Open to suggestions on that.

### CODE REVIEW

**Comment by tobymao** [comment]
can you add a test case for this? you can ensure that it raises

**Comment by georgesittas** [comment]
Did you actually come across cyclic scopes? How do you even reach that state? I'm somewhat skeptical about this due to the added complexity & overhead, not sure yet if it's justified.

File: tests/test_optimizer.py [hunk 1]
```diff
@@ -800,6 +800,13 @@ def test_scope_warning(self, logger):
             level="warning",
         )
 
+    def test_scope_circular_dependency(self):
+        scope = build_scope(parse_one("WITH w AS (SELECT * FROM q) SELECT * FROM w"))
+        cte_scope = scope.cte_scopes[0]
```
Comment (tobymao) [inline tests/test_optimizer.py:805]: this test doesn't make sense. why would this ever happen?

**Comment by hsheth2** [comment]
I've seen issues in the past where `build_scope` generated a cyclic setup. My original fix was to just set a max scope depth (https://github.com/tobymao/sqlglot/compare/main...hsheth2:sqlglot:main), but figured this would be a more reliable fix.

I'll try and to dig up the original SQL that caused it, but not sure I still have it handy.

**Comment by georgesittas** [comment]
My hunch is that this could only happen if an AST had different nodes with the same `id`, which would be a user issue to begin with, imo.

**Comment by tobymao** [comment]
i'm going to close this for now. if you can show a code path where this exists, i think we should fix that at parse time / optimization etc. if we cannot, then i'll accept a fix like this

**Comment by hsheth2** [comment]
Example script below - it will currently hang in `scope.traverse()` right now.

More broadly, I think that if we can detect situations where we might end up in an infinite loop, it makes sense to raise an error instead of simply letting it happen.

```python
import sqlglot
import sqlglot.optimizer

query = """\
SELECT
  clicks.user_id,
  clicks.email_address_id,
  sum(IF(clicks.utm_source = 'newsletter', 1, 0)) AS newsletter_clicks

FROM (
  SELECT emails.user_id, clicks.email_address_id, clicks.utm_source
  FROM `proj.campaign1.clicks` as clicks
  JOIN emails USING (email_address_id)
  WHERE clicks.utm_source IN (
    'a', 'b', 'c'
  )
  AND _date = current_run_date

  UNION ALL

  SELECT emails.user_id, clicks.email_address_id, clicks.utm_source
  FROM `proj.campaign2.clicks` as clicks
  JOIN emails USING (email_address_id)

  UNION ALL

  SELECT emails.user_id, clicks.email_address_id, clicks.utm_source
  FROM `proj.campaign3.clicks` as clicks
  JOIN emails USING (email_address_id)
) as clicks
"""

optimized = sqlglot.optimizer.optimizer.optimize(
    query,
    dialect="bigquery",
    validate_qualify_columns=False,
)
```

**Comment by tobymao** [comment]
@hsheth2 thanks for the example

**Comment by tobymao** [comment]
this is simply a bug in the optimizer, and the cyclic scope is only a patch to the underlying issue

**Comment by tobymao** [comment]
this is due to user error. downstream rules require full column qualification. since you passed in validate_qualify_columns=False, the rules were not correctly applied, causing unnest_subqueries to incorrectly detect that you have a decorrelated subquery

**Pull request closed without merging**